[dashboard] Implement Teams UI (selector, creation wizard, members page, project page)

jankeromnes opened the pull request:
Fixes https://github.com/gitpod-io/gitpod/issues/4355
Fixes #4354

TODO:
- [x] Update the top navigtation with tabs so that Workspaces and Settings are palced below the account dropdown
- [x] Move the Admin item to the right menu
- [x] Introduce a dropdown to switch between teams as well as create new teams
- [x] Introduce a full page form to create a team
- [x] Polish account dropdown to better align with spec
- [x] Toggle Teams & Projects UI on-and-off with a permission (thanks @svenefftinge for #4396!)
- [x] Surface basic membership information for Teams (e.g. user ID + full name)
- [x] A team is reachable by its URL which is `https://gitpod.io/<teamname>`
- [ ] Disallow team names that are also user names or that conflict with dashboard URLs (EDIT now tracked here: https://github.com/gitpod-io/gitpod/issues/4469)

jankeromnes commented:
Currently looks like this:

<img width="600" alt="Screenshot 2021-06-06 at 17 31 19" src="https://user-images.githubusercontent.com/599268/120930385-55904880-c6ed-11eb-90bb-7c490729fe1c.png">
<img width="600" alt="Screenshot 2021-06-06 at 17 35 53" src="https://user-images.githubusercontent.com/599268/120930481-adc74a80-c6ed-11eb-856d-c69d5705d8f7.png">


The new Teams UI can also be disabled (`showTeams=false`), making the top menu look almost exactly like before (except `Admin` is now on the right):

<img width="600" alt="Screenshot 2021-06-06 at 17 33 14" src="https://user-images.githubusercontent.com/599268/120930420-722c8080-c6ed-11eb-9940-87299c83df37.png">

svenefftinge commented:
Can you also include #4354 in this PR ([see](https://github.com/gitpod-io/gitpod/pull/4396#issuecomment-855644065))?

jankeromnes commented:
@svenefftinge Sure, many thanks!

jankeromnes commented:
🤣

```
  1) TeamDBSpec
       createAndFindATeam:
     Error: Team name should match /^[A-Za-z0-9]+$/
```

https://github.com/gitpod-io/gitpod/blob/f83bfadeebe840ee38d95ca5b50d2b30753079ca/components/gitpod-db/src/team-db.spec.db.ts#L45

jankeromnes commented:
Alright, this is now ready for review. 🎉

Please take a look! 🙏  In particular:

- @svenefftinge and @gtsiolis could you please review the overall UX?

- @AlexTugarev could you please review the addition of `teams` to the `UserContext`?

- @csweichel could you please review the `GuardedTeam` access guard?

svenefftinge commented on components/dashboard/src/Login.tsx:
```diff
@@ -62,8 +62,11 @@ export function Login() {
+            getGitpodService().server.getLoggedInUser(),
+            getGitpodService().server.getTeams(),
+        ]);
+        setContext({ user, teams });
```
I think we should rather have a Context per object. Both lifecycle and usages of user and teams seems too different.

svenefftinge commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -1385,11 +1386,36 @@ export class GitpodServerImpl<Client extends GitpodClient, Server extends Gitpod
+        const memberships = await this.teamDB.findMembershipsByTeam(team.id);
+        await this.guardAccess({ kind: "team", subject: team, memberships }, "get");
+        return Promise.all(memberships.map(async (m: DBTeamMembership): Promise<TeamMemberInfo> => {
+            const member = await this.userDB.findUserById(m.userId);
```
Depending on the team size this can become expensive. A single SQL query would be better.

jankeromnes replied:
https://github.com/gitpod-io/gitpod/issues/4470

svenefftinge reviewed (approved):
code looks good!

svenefftinge commented:
> Disallow team names that are also user names or that conflict with dashboard URLs

would be good to move this to an issue, so we don't forget :-)

gtsiolis commented:
/werft run

gtsiolis commented on components/dashboard/src/App.tsx:
```diff
@@ -148,6 +151,7 @@ function App() {
                 <Route path="/notifications" exact component={Notifications} />
                 <Route path="/plans" exact component={Plans} />
                 <Route path="/teams" exact component={Teams} />
+                <Route path="/new-team" exact component={NewTeam} />
```
**question**: Shall we use a more implicit path like `/teams/new` and also reserve this URL?

jankeromnes replied:
I'm not sure, because `/teams` is the old Team Plans page, while new team URLs look more like `/:slug/members`.

Let's leave it as is for now, and maybe adjust the URL later. 😇

jankeromnes replied:
Now tracked in https://github.com/gitpod-io/gitpod/issues/4471

gtsiolis commented on components/dashboard/src/components/Menu.tsx:
```diff
@@ -4,83 +4,191 @@
+                                }
+                            ]}>
+                                <div className="flex p-1.5 pl-3 rounded-lg hover:bg-gray-200">
+                                    <span className="text-base text-gray-600 font-semibold">{team?.name || userFullName}</span>
```
**issue**: Reminder to restore the colors for dark theme! These can use the colors of the menu items on the right. Font size can be `text-sm`! 🌙

jankeromnes replied:
@gtsiolis Unless I'm looking at the wrong spec, the [Figma file](https://www.figma.com/file/zU2ZmLwZzegEAHKwRH3xkL/Team-Members-%234421?node-id=0%3A1) currently has `text-base` (i.e. `16px`) for team member names, not `text-sm` (i.e. `14px`).

gtsiolis replied:
@jankeromnes You are absolutely right! The initial specs are using `16px` for the main and secondary navigation on the top right which is now still using a `14px` size. May I suggest the following two items:
1. Either use the smaller size of align all navigation elements and make sure they are using a `32px` height?
2. If we go with the specs (`16px`), semibold here is rendering quite bold so I'd either change the font weight to medium or even better enable font-smoothing for the product as in https://github.com/gitpod-io/website/pull/222.

jankeromnes replied:
Now tracked in https://github.com/gitpod-io/gitpod/issues/4471

gtsiolis commented on components/dashboard/src/components/Menu.tsx:
```diff
@@ -4,83 +4,191 @@
+                                ...(teams || []).map(t => ({
+                                    title: t.name,
+                                    customContent: <div className="w-full text-gray-400 flex flex-col">
+                                        <span className="text-gray-800 text-base font-semibold">{t.name}</span>
```
**issue**: Reminder to add colors for dark theme and use `text-sm` font size!

```suggestion
                                        <span className="text-gray-800 text-sm dark:text-gray-300 font-semibold">{t.name}</span>
```

jankeromnes replied:
@gtsiolis: The [spec](https://www.figma.com/file/24YZdNs7MmS5drlZ5ZlM7r/Allow-adding-and-switching-teams-%234355?node-id=617%3A179) currently has `text-base` a.k.a. `16px` as font size for team names in the team switcher-menu.

jankeromnes replied:
Now tracked in https://github.com/gitpod-io/gitpod/issues/4471

gtsiolis commented on components/dashboard/src/teams/NewTeam.tsx:
```diff
@@ -0,0 +1,49 @@
+        <h1>New Team</h1>
+        <p className="text-gray-500 text-center text-base">Teams allow you to <strong>group multiple projects</strong>, <strong>collaborate with others</strong>, <strong>manage subscriptions</strong> with one centralized billing, and more. <a href="">Learn more</a></p>
+        <form className="mt-8 w-full" onSubmit={createTeam}>
+            <div className="border rounded-xl p-6 border-gray-100">
```
```suggestion
            <div className="border rounded-xl p-6 border-gray-100 dark:border-gray-800">
```

gtsiolis commented on components/dashboard/src/components/Menu.tsx:
```diff
@@ -4,83 +4,191 @@
+                    </Link>
+                    <div className="ml-2 text-base">
+                        {showTeamsUI
+                            ? <ContextMenu classes="w-64 left-0" menuEntries={[
```
**question**: Could we `bg-gray-100` on hover and also use this background for the currently active team? 💭

jankeromnes replied:
Now tracked in https://github.com/gitpod-io/gitpod/issues/4471

gtsiolis commented on components/gitpod-db/src/typeorm/team-db-impl.ts:
```diff
@@ -28,10 +29,50 @@ export class TeamDBImpl implements TeamDB {
+            throw new Error('Team name cannot be empty');
+        }
+        if (!/^[A-Za-z0-9 _-]+$/.test(name)) {
+            throw new Error('Team name should match /^[A-Za-z0-9 _-]+$/');
```
**suggestion**: Could we use a friendlier error message when using special characters? Something like the following?

```suggestion
            throw new Error('Please choose a team name with no special characters.);
```

jankeromnes replied:
Good point! How about:

> Please choose a team name containing only letters, numbers, -, _, or spaces.

gtsiolis replied:
If I only I had this kind of explanation for every regex expression I ran into. 😅 

Sounds good, as long as we as remove that regex.

gtsiolis commented on components/dashboard/src/teams/NewTeam.tsx:
```diff
@@ -0,0 +1,49 @@
+                <h3 className="text-center text-xl mb-6">What's your team's name?</h3>
+                <h4>Team Name</h4>
+                <input className="w-full" type="text" onChange={event => name = event.target.value} />
+                {!!creationError && <p className="text-gitpod-red">{creationError.message}</p>}
```
**issue**: Ideally, this would also change the input border to red but should be fine to leave this as is for this iteration. We'll get back to input validation styles.

jankeromnes replied:
Added an `error` class for some inputs. Let's see how this looks. 😁

gtsiolis commented on components/dashboard/src/projects/Projects.tsx:
```diff
@@ -0,0 +1,23 @@
+        <Header title="Projects" subtitle="Manage recently added projects." />
+        <div>
+            <img alt="Projects (empty)" className="h-44 mt-24 mx-auto" role="presentation" src={projectsEmpty} />
+            <p className="text-center text-base text-gray-500 mt-16">Add projects to enable and manage Prebuilds.<br/><a className="underline" href="https://www.gitpod.io/docs/prebuilds/">Learn more about Prebuilds</a></p>
```
**issue**: Reminder to use the same style with `underline-thickness-thin` and `underline-offset-small` for all information links.

For example:

https://github.com/gitpod-io/gitpod/blob/c31e2f05fe3ccc7e9e1dbae0efa9454270b4026a/components/dashboard/src/workspaces/Workspaces.tsx#L124

jankeromnes replied:
Good point, thanks! Will create a `a.learn-more` class and re-use that everywhere.

gtsiolis commented on components/dashboard/src/projects/Projects.tsx:
```diff
@@ -0,0 +1,23 @@
+    return <>
+        <Header title="Projects" subtitle="Manage recently added projects." />
+        <div>
+            <img alt="Projects (empty)" className="h-44 mt-24 mx-auto" role="presentation" src={projectsEmpty} />
```
**praise**: This looks so good, that I added https://github.com/gitpod-io/gitpod/issues/4455. 😋

gtsiolis commented on components/dashboard/src/teams/NewTeam.tsx:
```diff
@@ -0,0 +1,49 @@
+    }
+    return <div className="flex flex-col w-96 mt-16 mx-auto items-center">
+        <h1>New Team</h1>
+        <p className="text-gray-500 text-center text-base">Teams allow you to <strong>group multiple projects</strong>, <strong>collaborate with others</strong>, <strong>manage subscriptions</strong> with one centralized billing, and more. <a href="">Learn more</a></p>
```
**issue**: Reminder to use the same style (underline) if possible. Alternatively, a blue color for the link could suffice for making this slightly more accessible.

gtsiolis commented on components/dashboard/src/teams/NewTeam.tsx:
```diff
@@ -0,0 +1,49 @@
+    return <div className="flex flex-col w-96 mt-16 mx-auto items-center">
+        <h1>New Team</h1>
+        <p className="text-gray-500 text-center text-base">Teams allow you to <strong>group multiple projects</strong>, <strong>collaborate with others</strong>, <strong>manage subscriptions</strong> with one centralized billing, and more. <a href="">Learn more</a></p>
+        <form className="mt-8 w-full" onSubmit={createTeam}>
```
**suggestion**: Let's give this some space!

```suggestion
        <form className="mt-16 w-full" onSubmit={createTeam}>
```

gtsiolis commented on components/dashboard/src/workspaces/StartWorkspaceModal.tsx:
```diff
@@ -47,8 +42,8 @@ export function StartWorkspaceModal(p: StartWorkspaceModalProps) {
-                <Tab name='Recent' setSelection={setSelection} selection={selection} />
-                <Tab name='Examples' setSelection={setSelection} selection={selection} />
+                <TabMenuItem name='Recent' selected={selection === 'Recent'} onClick={() => setSelection('Recent')} />
+                <TabMenuItem name='Examples' selected={selection === 'Examples'} onClick={() => setSelection('Examples')} />
```
**question**: Could we make this tab menu disappear when creating a new team similar with the specs? Low priority! 🔅

jankeromnes replied:
Now tracked in https://github.com/gitpod-io/gitpod/issues/4471

gtsiolis reviewed (approved):
Clicking those tabs [feels so good](https://youtu.be/D5Y11hwjMNs)! Thanks for adding this @jankeromnes! 🌟  

Left some first-round comments, feel free to open follow up issues for anything that can go in a later PR or ingore any comments that seem minor. 🏀

gtsiolis commented on components/dashboard/src/projects/Projects.tsx:
```diff
@@ -0,0 +1,23 @@
+        <Header title="Projects" subtitle="Manage recently added projects." />
+        <div>
+            <img alt="Projects (empty)" className="h-44 mt-24 mx-auto" role="presentation" src={projectsEmpty} />
+            <p className="text-center text-base text-gray-500 mt-16">Add projects to enable and manage Prebuilds.<br/><a className="underline" href="https://www.gitpod.io/docs/prebuilds/">Learn more about Prebuilds</a></p>
```
**issue**: Oh, this is missing a title element that I accidentally removed from the draft specs! You can find the latest specs for this page in https://github.com/gitpod-io/gitpod/issues/4420.

svenefftinge commented on components/server/src/user/user-service.ts:
```diff
@@ -147,6 +147,8 @@ export class UserService {
             newUser.rolesOrPermissions = ['admin'];
         }
+        // FIXME(janx): Remove this before deploying to production(!)
+        newUser.rolesOrPermissions = (newUser.rolesOrPermissions || []).concat(['teams-and-projects']);
```
instead of this, we could whitelist the preview env URLs in the menu.ts for now. I.e. `'teams-and-projects' or preview-env-url`

svenefftinge replied:
I mean here https://github.com/gitpod-io/gitpod/pull/4401/files#diff-64248da95f399a1cfe99b03aa662caf313faf9b7016999111605da5d61f1971dR43

jankeromnes replied:
Good idea, done. Also removed the WIP commit.

jankeromnes commented:
Thanks again for the fantastic reviews! ❤️

Also, _phew!_, addressed all nits (either here or by opening follow-up issues), and double/triple-checked everything again.

One more smoke test on this deployment (because I'm getting tired and had to do some pretty crazy interactive reverts/rebases) and we should be good to merge. 🚀
